Install our knative operators on minikube

jcrossley3 opened the pull request:
We can't use the Maistra operator, since it's dependent on OpenShift,
so we install the istio.yaml released with knative-serving instead.

We do install OLM and the knative operators, of course.

jcrossley3 committed:
Install our knative operators on minikube

We can't use the Maistra operator, since it's dependent on OpenShift,
so we install the istio.yaml released with knative-serving instead.

We do install OLM and the knative operators, of course.

kameshsampath commented on etc/scripts/install-on-minikube.sh:
```diff
@@ -0,0 +1,30 @@
+
+# configure knative profile
+minikube profile knative
+minikube config set kubernetes-version v1.11.5 -p knative
```
we dont need to set this things, the latest ones k8s v1.13.x works we well.  Also `-p` not required as you have set the profile earlier

jcrossley3 replied:
1.11.5 is upstream's recommendation. As for `-p`, I don't trust minikube's handling of profiles ;)

kameshsampath replied:
at least as of 0.34.x i see it consistent. anyways no big thing thought to avoid repeated things

kameshsampath commented on etc/scripts/install-on-minikube.sh:
```diff
@@ -0,0 +1,30 @@
+# configure knative profile
+minikube profile knative
+minikube config set kubernetes-version v1.11.5 -p knative
+minikube config set memory 10240 -p knative
```
this is bit of hard rule - i got a 16GB machine and if i give 10GB to minikube then am done :( .. Can ask user for it ?

kameshsampath replied:
Also we can remove `-p`

jcrossley3 replied:
The user will need a minimum of 8gb

kameshsampath replied:
yes lets set it as 8GB as default, since we are getting bigger machines today thought we could prompt the user

kameshsampath commented on etc/scripts/install-on-minikube.sh:
```diff
@@ -0,0 +1,30 @@
+minikube profile knative
+minikube config set kubernetes-version v1.11.5 -p knative
+minikube config set memory 10240 -p knative
+minikube config set cpus 4 -p knative
```
we can remove `-p`

kameshsampath commented on etc/scripts/install-on-minikube.sh:
```diff
@@ -0,0 +1,30 @@
+minikube config set kubernetes-version v1.11.5 -p knative
+minikube config set memory 10240 -p knative
+minikube config set cpus 4 -p knative
+minikube config set disk-size 50g -p knative
```
we can remove `-p`

kameshsampath commented on etc/scripts/install-on-minikube.sh:
```diff
@@ -0,0 +1,30 @@
+# Start minikube
+minikube start -p knative --extra-config=apiserver.enable-admission-plugins="LimitRanger,NamespaceExists,NamespaceLifecycle,ResourceQuota,ServiceAccount,DefaultStorageClass,MutatingAdmissionWebhook"
+
+#oc login -u admin -p admin
```
we can delete this?

jcrossley3 replied:
Yes

kameshsampath commented on etc/scripts/install.sh:
```diff
@@ -37,16 +39,22 @@ wait_for_all_pods knative-serving
+  oc adm policy add-scc-to-user anyuid -z default
+else
+  oc create namespace myproject
+  oc label namespace myproject istio-injection=enabled
```
this is not required with Knative's istio as that already has annotation based injection

kameshsampath replied:
I am also worried on the `oc` commands here. If the user just has `kubectl` then these ones will fail :(

jcrossley3 replied:
The label is from the upstream docs: https://github.com/knative/docs/blob/master/install/Knative-with-Minikube.md#installing-istio

I worry about `oc` vs `kubectl`, too. I need to isolate where only one will work. Off the top of my head, I'd need to know the `kubectl` equivalent of `oc whoami --show-server`

kameshsampath replied:
>The label is from the upstream docs: https://github.com/knative/docs/blob/master/install/Knative-with-Minikube.md#installing-istio

the upstream docs was not right, I did test it out as well without that

kameshsampath commented on etc/scripts/installation-functions.sh:
```diff
@@ -26,7 +26,17 @@ function wait_for_all_pods {
+function check_minikube {
+  (hash minikube &&
+     minikube ip | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b" &&
+     oc whoami --show-server | grep "$(minikube ip)"
```
no oc for minkube

jcrossley3 replied:
What's the equivalent for `kubectl`?

jcrossley3 replied:
Ok, I think I figured it out. With the latest push, you should be able to have either `kubectl` or `oc` installed for the script to work.

kameshsampath commented on etc/scripts/installation-functions.sh:
```diff
@@ -40,6 +50,8 @@ function check_operatorgroups {
   if check_openshift_4; then
     echo "Detected OpenShift 4 - skipping enabling admission webhooks."
+  elif check_minikube; then
+    echo "Detected minikube - skipping enabling admission webhooks."
```
Skipping enabling admission webhook is slightly misleading as we are enabling it via --extra-config

kameshsampath commented on etc/scripts/installation-functions.sh:
```diff
@@ -150,18 +162,23 @@ function install_olm {
+  if check_minikube; then
+    echo "Detected minikube - incompatible with Maistra operator, so installing upstream istio."
+    oc apply -f "https://github.com/knative/serving/releases/download/${KNATIVE_SERVING_VERSION}/istio.yaml"
+    oc label namespace default istio-injection=enabled
```
same oc comment as said earlier

jcrossley3 committed:
Incorporating PR feedback

Specifically adding configuration parameters with reasonable defaults,
e.g.
  $ MEMORY=10240 DISK_SIZE=30g ./etc/scripts/install-on-minikube.sh

jcrossley3 committed:
Use either kubectl or oc, whichever is available

But if targeting OpenShift, oc needs to be available for things like
`oc adm ...` and `oc login ...`. Fortunately, minikube doesn't require
those things in order to install knative via OLM.

kameshsampath commented on etc/scripts/installation-functions.sh:
```diff
@@ -136,32 +161,36 @@ function install_olm {
-    cat <<-EOF | oc apply -f -
+  if check_minikube; then
+    echo "Detected minikube - incompatible with Maistra operator, so installing upstream istio."
+    $CMD apply -f "https://github.com/knative/serving/releases/download/${KNATIVE_SERVING_VERSION}/istio.yaml"
```
@jcrossley3  can we add the crds' install to avoid initial warning/errors that user might see because of the missing crds ? 
`$CMD kubectl apply --filename https://github.com/knative/serving/releases/download/v0.3.0/istio-crds.yaml`

jcrossley3 replied:
Good call!

kameshsampath reviewed (commented):
Thanks for incorporating the feedback. It looks very nice now

jcrossley3 committed:
Fix quotes and install istio CRD's separately
